core/backend/src/rpc-impl/IModelReadRpcImpl.ts

```diff
@@ -194,6 +194,37 @@ export class IModelReadRpcImpl extends RpcInterface implements IModelReadRpcInte
     return iModelDb.getMassProperties(props);
   }
 
+  public async getMassPropertiesPerCandidate(tokenProps: IModelRpcProps, props: MassPropertiesPerCandidateRequestProps): Promise<MassPropertiesPerCandidateResponseProps[]> {
+    const iModelDb = await RpcBriefcaseUtility.findOpenIModel(RpcTrace.expectCurrentActivity.accessToken, tokenProps);
+    const result: MassPropertiesPerCandidateResponseProps[] = [];
+
+    for (const candidate of CompressedId64Set.iterable(props.candidates)) {
+      const massPropResults: MassPropertiesResponseProps[] = [];
+
+      for (const op of props.operations) {
+        const massProperties = await iModelDb.getMassProperties({ operation: op, candidates: [candidate] });
```

Comment:
getMassProperties executes on a thread pool, but you are serializing the processing. Consider switching to `Promise.allSettled`, and measure effect on your timings.

Reply:
> One thing that I noticed throughout all performance test runs is that memory usage of the backend is growing steadily.

If Parasolid is being used this isn't unexpected, it has its own caches...do you know if you were requesting mass properties for brep geometry?

Reply:
I just realized that all of that testing was done on elements without any geometry (GeometryStream = null). I don't think Parasolid would be used if there is no geometry.

I'll rerun the tests on elements with GeometryStream != null. The iModel that I am using for testing doesn't look like it has any weird geometry that would require BReps. Visually all of it looks like it could be modeled with simple extrusions. But it's also created using MicroStation, so it might be BReps. Is there some kind of function in itwinjs to check the geometry type of an element or geometry stream?

Reply:
Results from retesting using elements that have a valid GeometryStream:

| | Impl. 1 total | Impl. 1 avg | Impl. 2 total | Impl. 2 avg | Impl. 3 total | Impl. 3 avg |
| -- | -- | -- | -- | -- | -- | -- |
Batch 1 (1000) | 80600ms | 0.400ms | 40090 | 0.199ms | 30730ms | 0.152ms |
Batch 2 (1000) | 90978ms | 0.452ms | 72396ms | 0.360ms | 47495ms | 0.236ms |
Batch 3 (1000) | 102704ms | 0.510ms | 99049ms | 0.492ms | 58480ms | 0.290ms |
Batch 4 (438) | 50173ms | 0.569ms | 45524ms | 0.517ms | 29838ms | 0.338ms |
Batch 5 (1000) | 119810ms | 0.596ms | 145702ms | 0.724ms | 79733ms | 0.396ms |
Batch 6 (1000) | 133965ms | 0.666ms | 173969ms | 0.865ms | 86471ms | 0.430ms |
Batch 7 (1000) | 144139ms | 0.717ms | 211290ms | 1.051ms | 85135ms | 0.423ms |
Batch 8 (438) | 69374ms | 0.788ms | 88063ms | 1.000ms | 39181ms | 0.445ms |
Batch 9 (1000) | 159587ms | 0.793ms | 247082ms | 1.229ms | 99939ms | 0.497ms |
Batch 10 (1000) | 173499ms | 0.863ms | 264146ms | 1.314ms | 101281ms | 0.503ms |
Batch 11 (1000) | 183765ms | 0.914ms | 308708ms | 1.535ms | 102787ms | 0.511ms |
Batch 12 (438) | 87325ms | 0.991ms | 108257ms | 1.229ms | 46088ms | 0.523ms |

Will change to Implementation 3 - only the outer loop is rewritten to create promises which are later awaited using `Promise.all`.

Reply:
Changed to use `Promise.all`.

Reply:
Please do not resolve other people's comments on your PRs.
`Promise.all` rejects immediately if any of the promises rejects. Is that what you want?

Reply:
Sorry about the resolved comments.

Ideally, none of the promises in this method should ever be rejected, but rather resolve with `status === BentleyStatus.ERROR`. If any promise is rejected then the candidate for which it failed is lost and that would require iterating over all the candidates again to find which of them were rejected.

Also: `src/rpc-impl/IModelReadRpcImpl.ts(226,43): error TS2550: Property 'allSettled' does not exist on type 'PromiseConstructor'. Do you need to change your target library? Try changing the 'lib' compiler option to 'es2020' or later.`

@pmconne should I add a try/catch to the inner promises to always resolve with `BentleyStatus.ERROR` if any error occurs?

@bbastings does `getMassProperties` ever throw exceptions? Does it ever return status which is not `BentleyStatus.SUCCESS` or `BentleyStatus.ERROR`?

Reply:
The only status values are SUCCESS or ERROR.

Reply:
@bbastings is there a try/catch in `getMassProperties` that would catch all Parasolid errors and translate those to `BentleyStatus.ERROR`? Or is there a way for some errors to slip through as exceptions?

Reply:
I believe you will get an exception if Parasolid throws a severe error, there is no try/catch in getMassProperties.